src/hotspot/share/oops/instanceKlass.cpp

```diff
@@ -211,31 +211,39 @@ bool InstanceKlass::has_nest_member(JavaThread* current, InstanceKlass* k) const
   return false;
 }
 
-// Called to verify that k is a permitted subclass of this class
-bool InstanceKlass::has_as_permitted_subclass(const InstanceKlass* k) const {
+// Called to verify that k is a permitted subclass of this class.
+// The incoming stringStream is used to format the messages for error logging and for the caller
+// to use for exception throwing.
+bool InstanceKlass::has_as_permitted_subclass(const InstanceKlass* k, stringStream& ss) const {
   Thread* current = Thread::current();
   assert(k != nullptr, "sanity check");
   assert(_permitted_subclasses != nullptr && _permitted_subclasses != Universe::the_empty_short_array(),
          "unexpected empty _permitted_subclasses array");
 
   if (log_is_enabled(Trace, class, sealed)) {
     ResourceMark rm(current);
-    log_trace(class, sealed)("Checking for permitted subclass of %s in %s",
+    log_trace(class, sealed)("Checking for permitted subclass %s in %s",
                              k->external_name(), this->external_name());
   }
 
   // Check that the class and its super are in the same module.
   if (k->module() != this->module()) {
-    ResourceMark rm(current);
-    log_trace(class, sealed)("Check failed for same module of permitted subclass %s and sealed class %s",
-                             k->external_name(), this->external_name());
+    ss.print("Failed same module check: subclass %s is in module \"%s\" for loader %s "
+             "and sealed class %s is in module \"%s\" for loader %s",
+             k->external_name(), k->module()->name_as_C_string(), k->module()->loader_data()->loader_name_and_id(),
+             this->external_name(), this->module()->name_as_C_string(), this->module()->loader_data()->loader_name_and_id());
+    log_trace(class, sealed)(" - %s", ss.as_string());
     return false;
   }
 
   if (!k->is_public() && !is_same_class_package(k)) {
-    ResourceMark rm(current);
-    log_trace(class, sealed)("Check failed, subclass %s not public and not in the same package as sealed class %s",
-                             k->external_name(), this->external_name());
+    ss.print("Failed same run-time package check: non-public subclass %s is in package \"%s\" with classloader %s,"
```

Comment:
Since we paste the same module check at this point then it might be simpler to say that it fails the same package check.

Reply:
Ah I see what you are saying. As they are known to be in the same module then they are known to have the same loader. But "runtime package" is only significant for different loaders with the same package name - which can't be the case here. So it is simply not the same package. Will fix.

Thanks for looking at this @AlanBateman !